storage/provider.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"flag"
+	"fmt"
+	"sync"
+
+	"github.com/google/trillian/monitoring"
+)
+
+// NewProviderFunc is the signature of a function which can be registered to
+// provide instances of storage providers.
+type NewProviderFunc func(monitoring.MetricFactory) (Provider, error)
+
+var (
+	// TODO(pavelkalinnikov): Move this flag to main file.
+	storageSystem = flag.String("storage_system", "mysql", fmt.Sprintf("Storage system to use. One of: %v", providers()))
+
+	spMu     sync.RWMutex
+	spOnce   sync.Once
+	spByName map[string]NewProviderFunc
+)
+
+// RegisterProvider registers the given storage Provider.
+func RegisterProvider(name string, sp NewProviderFunc) error {
+	spMu.Lock()
+	defer spMu.Unlock()
+
+	spOnce.Do(func() {
+		spByName = make(map[string]NewProviderFunc)
```

Comment:
I know this is how the code already was, so you shouldn't change this in this PR, but I was wondering why `spByName` is initialized in this way, rather than just having `spByName = make(map[string]NewProviderFunc)` in the var block at top of the file. Any idea?

Reply:
No idea, looks like a poor man's optimization to me. Refactored it.